graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java

```diff
@@ -0,0 +1,143 @@
+        this.scopedEntityMongoUtils = mongoCollections.scopedEntityUtils(collection, entityScopeService);
+        this.paginationHelper = mongoCollections.paginationHelper(collection);
+        this.searchQueryParser = new SearchQueryParser(EntityGroup.FIELD_ENTITIES, ALLOWED_FIELDS);
+
+        final IndexOptions caseInsensitiveOptions = new IndexOptions()
+                .collation(Collation.builder().locale("en").collationStrength(CollationStrength.SECONDARY).build())
+                .unique(true);
+        collection.createIndex(new BasicDBObject(EntityGroup.FIELD_NAME, 1), caseInsensitiveOptions);
+    }
+
+    public Optional<EntityGroup> get(String id) {
+        return mongoUtils.getById(id);
+    }
+
+    public PaginatedList<EntityGroup> findPaginated(String query, int page, int perPage, Bson sort, Predicate<EntityGroup> filter) {
+        final SearchQuery searchQuery = searchQueryParser.parse(query);
+        return filter == null ?
+                paginationHelper.filter(searchQuery.toBson()).sort(sort).perPage(perPage).page(page) :
+                paginationHelper.filter(searchQuery.toBson()).sort(sort).perPage(perPage).page(page, filter);
+    }
+
+    public EntityGroup save(EntityGroup entityGroup) {
+        if (entityGroup.id() != null) {
+            return scopedEntityMongoUtils.update(entityGroup);
+        }
+        String newId = scopedEntityMongoUtils.create(entityGroup);
+        return entityGroup.toBuilder().id(newId).build();
+    }
+
+    public Optional<EntityGroup> getByName(String name) {
+        final Bson query = eq(EntityGroup.FIELD_NAME, name);
+
+        return Optional.ofNullable(collection.find(query).first());
+    }
+
+    public List<EntityGroup> getAllForEntity(String type, String entityId) {
+        final Bson query = and(
+                exists(typeField(type)),
+                in(typeField(type), entityId)
+        );
+        return MongoUtils.stream(collection.find(query)).toList();
+    }
+
+    // TODO: can we make this better..?
+    public Map<String, Collection<EntityGroup>> getAllForEntities(String type, Collection<String> entityIds) {
+        final Bson query = and(
+                exists(typeField(type)),
+                in(typeField(type), entityIds)
+        );
+        final List<EntityGroup> groups = MongoUtils.stream(collection.find(query)).toList();
+        final Multimap<String, EntityGroup> entityToGroupsMap = MultimapBuilder.hashKeys().hashSetValues().build();
+        for (EntityGroup group : groups) {
+            for (Map.Entry<String, Set<String>> StringEntry : group.entities().entrySet()) {
+                if (StringEntry.getKey().equals(type)) {
+                    for (String entityId : StringEntry.getValue()) {
+                        if (entityIds.contains(entityId)) {
+                            entityToGroupsMap.put(entityId, group);
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Some comments:
* `group.entities()` is nullable, so it should be checked, however in practice the query guarantees that the group will have entities. If that's always the case on the DB level, maybe reconsider the `@Nullable` annotation on `entities()`.
* Loop over `entities().entrySet()` and `if (StringEntry.getKey().equals(type)) {` should be a simple `get(type)`? The query already makes sure all groups contain something from that type, we just don't know which entityId it is.
* The overall code would benefit from performing this as an aggregation query in MongoDB, instead of doing it in code.